### PR TITLE
[JENKINS-29976] Create an attribute to set Dockerfile in @DockerFixture 

### PR DESCRIPTION
**Pull request by lordofthejars**
Create an attribute to set Dockerfile in @DockerFixture. This PR also fixes one small issue [JENKINS-29984](https://issues.jenkins-ci.org/browse/JENKINS-29984) but since it was found while writing a test for [JENKINS-29976](https://issues.jenkins-ci.org/browse/JENKINS-29976) it was difficult to split at that stage to two different commits because the bug was at the same class that I touched for implementing this feature. Since it is an small bug fix (support for inner classes for DockerFixture), I decided to not overcomplicate the development by removing/copyback the code.

I paste here the description of the issue:

```
Create attribute to set Dockerfile location in @DockerFixture. By default Dockerfile fixture should be placed in the resource directory and at the same package as DockerContainer subtype.

A new optional attribute called dockerfile can be added at @DockerFixture, so if it is not present then default behaviour is the expected, but if the attribute is present, Dockerfile location is get from there. This allows save Dockerfile in places where they can be reused.

An example of the annotation might be:

@DockerFixture(id="jira", ports="2909", dockerfile="a/b")
public class JiraContainer ....

Classpath resolution will be based on class where annotation is defined.
```

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 1]
```diff
@@ -158,6 +125,107 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
         }
     }
 
+    //package scope for testing purposes. Ideally we should encapsulate Docker interactions so they can be mocked
+    // and call public method.
+    void copyDockerfileDirectory(Class<? extends DockerContainer> fixture, DockerFixture f, File dir)
+            throws IOException {
+        String dockerfilePath = resolveDockerfileLocation(fixture, f);
+
+        if(dockerfilePath.startsWith(FILE_SCHEMA)) {
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: ```FileUtils.copyURLToFile(URL source, File destination)``` could be simpler to use, and you would get support for http e.al. at the same time.

Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: I thought the same but then I open the door that someone could use any schema like `http` or `ftp` since Java provides out of the box this support. So I decided to put some protection on this.
But of course I could implement the `if`as now but then call `copyURLToFile` as you mention.

**Comment by jglick** [comment]
I see no purpose for the file schema. Just get rid of the protocol and treat the value as a resource name. Simplifies the code and nothing of value is lost.

**Comment by lordofthejars** [comment]
I was thinking on more global purpose where users are not limited only in
classpath location and also opening to other protocols, but let s simplify
the code and if it is required in future, we can add it.
El dt., 18 d’ag., 2015 a les 18.22 Jesse Glick <notifications@github.com>
va escriure:

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 2]
```diff
@@ -158,6 +122,80 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
...
+        }
+        return prefix;
+    }
+
+    private void copyDockerfileDirectoryFromClasspath(Class<? extends DockerContainer> fixture, String dockerfileLocation, File dir) throws IOException {
+        File jar = null;
+        try {
+            jar = Which.jarFile(fixture);
+        } catch (IllegalArgumentException e) {
+            // fall through
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Some logging would be useful

File: src/test/java/org/jenkinsci/test/acceptance/docker/DockerTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Map;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+/**
+ * @author asotobueno
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/test/acceptance/docker/DockerTest.java]: NIT: If you add the Javadoc header, it's useful to add some info. E.g. <code>Tests for {@link DockerFixture}</code>. Probably the class could be also renamed, because actually we test the fixtures.

**Comment by oleg-nenashev** [comment]
Looks good to me in general, only minor comments

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 2 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: yes I completely agree but this was originally there, and since it is an improvement I didn't want to touch this as side effect. I have seen this structure in more places so we would need to create an issue for fixing some of them. Also in this case is not bad at all (maybe we could log as info) since if/else that are below this line takes care of everything. But as I said I agree with you that we need to log these lines. 

Do you think it has sense to open another issue with these improvements?

File: src/test/java/org/jenkinsci/test/acceptance/docker/DockerTest.java (hunk 3 above)
Comment (lordofthejars) [inline src/test/java/org/jenkinsci/test/acceptance/docker/DockerTest.java]: Absolutely.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 2 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: What logger framework you usually use? I tend to use JUL.

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: I don't think it's really required for the internal test framework. Since you're decoupling the code here, you can just add such messaging as a "refactoring". I would just keep it as a separate commit just to highlight the change for developers if they are interested in.

Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Yes in separate commit/PR make sense to not mix things. To log it you mean using Sysout directly?

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: probably a default java.logging.Logger would be the best option, but I'm not sure which log listeners are attached at this point.

Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: yes I tend to use java.logging.Logger (JUL) in this situations to not add a new framework

**Comment by olivergondza** [comment]
Is there any use-case for reusing dockerfile where you can not result whole fixture?

**Comment by lordofthejars** [comment]
Not now since it copies the whole directory. For what I understand you mean copying only Dockerfile and not the directory and the contents where it is stored isn't it? If it is the case, it is not implemented but we could open a new issue to do it.

**Comment by olivergondza** [comment]
I guess I do not understand what are you trying to achieve. My point was, why sharing dockerfiles between fixtures if you can reuse those fixtures for multiple tests?

**Comment by lordofthejars** [comment]
Because by default a Dockerfile must be in the same package and class name of the DockerFixture. What I am trying to achieve is breaking this hard dependency and allow users to put fixtures where they think it is better for the test. For example I have two test classes (not test methods) that wants to share same Dockerfile, then you can do it. But what it is really important this change is because it is a prior step to https://issues.jenkins-ci.org/browse/JENKINS-30009 and be able to make DockerFixture classes self-contained.

**Comment by olivergondza** [comment]
Well, I think of Dockerfile as an implementation detail of its fixture (which seems to be useful for JENKINS-30009). This change exposes Dockerfiles for reuse and encourages other fixtures to depend on your dockerfile content which seem to be the exact opposite of what you want.

Why can you not create *one* fixture and use it in *both* your tests?

**Comment by lordofthejars** [comment]
For what I see, in previous implementation you, before this PR, directory is automatically resolved. So if you have `public class ATest` and `public class Btest` then the directories where the fixtures are loaded are `<package>/ATest` and `<package>/BTest` so it is impossible to reuse them.

**Comment by olivergondza** [comment]
As I said `SshdContainer` is shared between `PublishOverSSHPluginTest` and `ScpPluginTest` so there is a way to reuse. In what regard it does not work for you?

As we failed to understand each other until this point, do you mind implementing your fixtures with duplicity and refactor the fixture sharing after that to demonstrate what the problem actually is?

**Comment by lordofthejars** [comment]
Sure you can do it in this way, and you can still do it in this way, but what I am implementing here is a way to allow developer to add fixtures in different directory and not depending only in defaults. The old behaviour is still valid and working so I simply enrich a bit more the project with this feature which adds more freedom to developers. It is not a problem itself with the current implementation it is simply an enhancement to the project.

**Comment by olivergondza** [comment]
I see. I am -0 on adding more ways to do the same thing.

**Comment by lordofthejars** [comment]
It is not the same thing, for example if I want to have all fixtures in src/test/resources/fixtures, then it is impossible. This is something that all test frameworks that works with fixtures offers. I don't see any problem.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 4]
```diff
@@ -158,6 +122,80 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
         }
     }
 
+    //package scope for testing purposes. Ideally we should encapsulate Docker interactions so they can be mocked
+    // and call public method.
+    void copyDockerfileDirectory(Class<? extends DockerContainer> fixture, DockerFixture f, File dir)
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Seems like `@Nonnull` etc might be appropriate?

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 5]
```diff
@@ -158,6 +122,80 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
...
+    }
+
+    private String resolveDockerfileLocation(Class<? extends DockerContainer> fixture, DockerFixture f) {
+        String prefix = null;
+        if(isSpecificDockerfileLocationSet(f)) {
+            prefix = f.dockerfile().replace('.', '/').replace('$', '/');
+        } else {
+            prefix = fixture.getName().replace('.', '/').replace('$', '/');
+        }
+        return prefix;
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Might be clearer to do the replace in one place?

E.g.

```java
String prefix = null;
if(isSpecificDockerfileLocationSet(f)) {
    prefix = f.dockerfile();
} else {
    prefix = fixture.getName();
}
return prefix.replace('.', '/').replace('$', '/');
```

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 6]
```diff
@@ -158,6 +122,80 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
...
+        return f.dockerfile() != null && !"".equals(f.dockerfile().trim());
+    }
+
+    private void copyDockerfileDirectoryFromPackaged(File jar, String fixtureLocation, File outputDirectory) throws IOException {
+        try (JarFile j = new JarFile(jar)) {
+            Enumeration<JarEntry> e = j.entries();
+            while (e.hasMoreElements()) {
+                JarEntry je = e.nextElement();
+                if (je.getName().startsWith(fixtureLocation)) {
+                    File dst = new File(outputDirectory, je.getName().substring(fixtureLocation.length()));
```
Comment (tfennelly) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: possible :bug:  ... wouldn't you need to check the length here before doing `substring` ?

Comment (tfennelly) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Sorry ... ignore ... I see you did a `startsWith` check before it.

**Comment by tfennelly** [comment]
:bee:

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 5 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: True, I am going to refactor :)

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 4 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Well I am not aware of the politics you follow regarding this, in package scope you also use them?

File: docs/FIXTURES.md [hunk 7]
```diff
@@ -63,4 +63,20 @@ The `@DockerFixture` bindIp allows you to bind the accessible ports of the docke
...
 `org/acme/FooContainer.java`, then the docker file must be located at `org/acme/FooContainer/Dockerfile`.
 
+In case of inner classes, dollar sign ( $ ) is replaced with a slash ( / ).
+
+You can also set custom location of `Dockerfile` by using `dockerfile` attribute from`@DockerFixture` annotation.
+A simple example might be:
+
+```java
+@DockerFixture(id="test", ports="8080", dockerfile="org/acme/fixture/test")
+publicclass TestContainer extends DockerContainer {
```
Comment (jglick) [inline docs/FIXTURES.md]: missing space

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java [hunk 8]
```diff
@@ -38,4 +39,23 @@
...
+     *  By default Dockerfile fixture should be placed in the resource directory
+     *  and at the same package as DockerContainer subtype.
+     * </p>
+     * <p>
+     *  If this attribute is present, Dockerfile file specified in the attribute
+     *  is used as Dockerfile fixture.
+     * </p>
+     * <p>
+     *  The location can be expressed in terms of file location (file:///a/b/Dockerfile)
+     *  or classpath location (classpath://a/b/Dockerfile).
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: I thought you removed this feature?

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java [hunk 9]
```diff
@@ -38,4 +39,23 @@
...
+    /**
+     * Path of Dockerfile file.
+     *
+     * <p>
+     *  By default Dockerfile fixture should be placed in the resource directory
+     *  and at the same package as DockerContainer subtype.
+     * </p>
+     * <p>
+     *  If this attribute is present, Dockerfile file specified in the attribute
+     *  is used as Dockerfile fixture.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: According to the documentation and the tests, this attribute specifies the (resource) path to the _folder_ containing `Dockerfile`, not the `Dockerfile` itself. So this Javadoc is misleading, as is the name of the attribute.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 4 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Generally we like to add nullness annotations everywhere we can.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 10]
```diff
@@ -158,6 +122,80 @@ public DockerImage build(Class<? extends DockerContainer> fixture) throws IOExce
...
+            dockerfileLocation += "/";
+            copyDockerfileDirectoryFromPackaged(jar, dockerfileLocation, dir);
+        } else {
+            // Dockerfile is not packaged into a jar file, so copy locally
+            copyDockerfileDirectoryFromLocal(dockerfileLocation, dir);
+        }
+    }
+
+    private boolean isSpecificDockerfileLocationSet(DockerFixture f) {
+        return f.dockerfile() != null && !"".equals(f.dockerfile().trim());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Annotation values cannot be null,

```
@DockerFixture(…, dockerfile=" ")
```

seems like an unlikely user mistake, and comparing to `""` is deprecated. This method could just be inlined as

```java
!f.dockerfile().isEmpty()
```

**Comment by jglick** [comment]
The added tests are good. As with @olivergondza I see no real motivation for the change and minor added API complexity—the conventional location worked fine—but it seems harmless enough.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java (hunk 8 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: Ups I updated doc but not javadoc.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java (hunk 9 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: Maybe `dockerfileFolder` may be better? I am don't know exactly how to name it

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 10 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Yes I know but I have find in  my career doing such "awesome" things using reflection. But well let's assume that in this case since it is for testing we won't get this things.

**Comment by lordofthejars** [comment]
Ok, then the maintainer can accept or close the PR, but this conversion goes to nowhere. As I don't see any complication on the API, in fact I have refactored a lot of things that originally were not refactored and makes difficult to follow, and simply I have added new features to the API.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 4 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Ok I have seen that in this project it is using java8 annotations for non nullable.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java (hunk 9 above)
Comment (lordofthejars) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerFixture.java]: Since I agree with you that `dockerfile` doesn't describe very well the purpose I have changed to `dockerFolder`

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 11]
```diff
@@ -27,6 +27,7 @@
  * Use this subsystem by injecting this class into your test.
  *
  * @author Kohsuke Kawaguchi
+ * @author asotobueno
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: BTW you can also feel free (IMO) to just delete any `@author` tags. Anyone who cares can `git blame` for themselves.

**Comment by jglick** [comment]
:bee:
